sources/sqlserver/toddl.go

```diff
@@ -0,0 +1,103 @@
+// mapping.  toSpannerType returns the Spanner type and a list of type
+// conversion issues encountered.
+func (tdi ToDdlImpl) ToSpannerType(conv *internal.Conv, columnType schema.Type) (ddl.Type, []internal.SchemaIssue) {
+	ty, issues := toSpannerTypeInternal(conv, columnType.Name, columnType.Mods)
+	if conv.TargetDb == constants.TargetExperimentalPostgres {
+		ty = overrideExperimentalType(columnType, ty)
+	} else {
+		if len(columnType.ArrayBounds) > 1 {
+			ty = ddl.Type{Name: ddl.String, Len: ddl.MaxLength}
+			issues = append(issues, internal.MultiDimensionalArray)
+		}
+		ty.IsArray = len(columnType.ArrayBounds) == 1
+	}
+	return ty, issues
+}
+
+// toSpannerType maps a scalar source schema type (defined by id and
+// mods) into a Spanner type. This is the core source-to-Spanner type
+// mapping.  toSpannerType returns the Spanner type and a list of type
+// conversion issues encountered.
+func toSpannerTypeInternal(conv *internal.Conv, id string, mods []int64) (ddl.Type, []internal.SchemaIssue) {
+	switch id {
+	case "bit":
+		return ddl.Type{Name: ddl.Bool}, nil
+	case "uniqueidentifier":
+		return ddl.Type{Name: ddl.String, Len: ddl.MaxLength}, nil
+	case "binary", "varbinary", "image":
+		if len(mods) > 0 && mods[0] > 0 && mods[0] <= ByteLimit {
+			return ddl.Type{Name: ddl.Bytes, Len: mods[0]}, nil
+		}
+		return ddl.Type{Name: ddl.Bytes, Len: ddl.MaxLength}, nil
+	case "date":
+		return ddl.Type{Name: ddl.Date}, nil
+	case "float", "real":
+		return ddl.Type{Name: ddl.Float64}, []internal.SchemaIssue{internal.Widened}
+	case "bigint":
+		return ddl.Type{Name: ddl.Int64}, nil
+	case "tinyint", "smallint", "int":
+		return ddl.Type{Name: ddl.Int64}, []internal.SchemaIssue{internal.Widened}
+	case "numeric", "money", "smallmoney", "decimal":
+		return ddl.Type{Name: ddl.Numeric}, nil
+	case "ntext", "text", "xml":
+		return ddl.Type{Name: ddl.String, Len: ddl.MaxLength}, nil
+	case "smalldatetime", "datetimeoffset", "datetime2", "datetime", "timestamp":
+		return ddl.Type{Name: ddl.Timestamp}, []internal.SchemaIssue{internal.Timestamp}
+	case "time":
+		return ddl.Type{Name: ddl.String, Len: ddl.MaxLength}, []internal.SchemaIssue{internal.Time}
+	case "varchar", "char", "nvarchar", "nchar":
+		if len(mods) > 0 && mods[0] > 0 && mods[0] <= StringLimit {
+			return ddl.Type{Name: ddl.String, Len: mods[0]}, nil
+		}
+		return ddl.Type{Name: ddl.String, Len: ddl.MaxLength}, nil
+	default:
+		return ddl.Type{Name: ddl.String, Len: ddl.MaxLength}, []internal.SchemaIssue{internal.NoGoodType}
+	}
+}
+
+// Override the types to map to experimental postgres types.
+func overrideExperimentalType(columnType schema.Type, originalType ddl.Type) ddl.Type {
+	if len(columnType.ArrayBounds) > 0 {
+		return ddl.Type{Name: ddl.String, Len: ddl.MaxLength}
+	} else if columnType.Name == "date" {
+		return ddl.Type{Name: ddl.String, Len: ddl.MaxLength}
+	} else if columnType.Name == "json" || columnType.Name == "jsonb" {
```

Comment:
I think sql server does noth ave a json tpe so this can be removed.

Reply:
agreed, will be changed in the next commit.